File: tests/e2e/fixtures/code/sample.lua
-- Lua (legacy stream mode): tables, closures, string methods.
local Note = {}
Note.__index = Note

function Note.new(path, tags)
  return setmetatable({ path = path, tags = tags or {} }, Note)
end

function Note:size()
  local f = io.open(self.path, "rb")
  if not f then return 0 end
  local n = f:seek("end")
  f:close()
  return n
end

local function groupByTag(notes)
  local out = {}
  for _, note in ipairs(notes) do
    for _, tag in ipairs(note.tags) do
      out[tag] = out[tag] or {}
      table.insert(out[tag], note)
    end
  end
  print(("grouped %d notes"):format(#notes))
  return out
end

return { Note = Note, groupByTag = groupByTag }

File: tests/e2e/fixtures/code/sample.kt
// Kotlin (legacy stream mode): data classes, extension functions, when.
import java.io.File

data class Note(val path: String, val tags: List<String> = emptyList()) {
    val size: Long get() = File(path).length()
}

fun List<Note>.groupByTag(): Map<String, List<Note>> =
    flatMap { n -> n.tags.map { it to n } }
        .groupBy({ it.first }, { it.second })

fun describe(n: Int) = when {
    n == 0 -> "no tags"
    n > 100 -> "many tags: $n"
    else -> "$n tags"
}

fun main() {
    val notes = listOf(Note("a.md", listOf("x")), Note("b.md"))
    println(describe(notes.groupByTag().size))
}
